subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassSetAnalysis.java

```diff
@@ -63,19 +63,26 @@ public DependentsSet getRelevantDependents(String className, Set<Integer> consta
         if (deps == null && constants.isEmpty()) {
             return DefaultDependentsSet.EMPTY;
         }
+        // since 3.4.1, the Set<Integer> of constants here doesn't match the set of literals, so we cannot use them
+        // anymore. If the set above is not empty, it means a constant has been added or changed, and we need to
+        // recompile everything
+        if (!constants.isEmpty()) {
+            return DependencyToAll.INSTANCE;
+        }
         Set<String> result = new HashSet<String>();
         if (deps != null && !deps.isDependencyToAll()) {
             recurseDependents(new HashSet<String>(), result, deps.getDependentClasses());
         }
-        for (Integer constant : constants) {
+        /*for (Integer constant : constants) {
```

Comment:
Okay... I kept it explicitly because it's easier to figure out why there's a `Set<Integer>` as a parameter. You could argue we have `git log` but sometimes I find it easier to keep the code to describe why it's not there anymore.